restli-client/src/main/java/com/linkedin/restli/client/ParSeqBasedCompletionStage.java

```diff
@@ -0,0 +1,621 @@
+        }
+      });
+      return promise;
+    })).toCompletionStage();
+  }
+
+  @Override
+  public ParSeqBasedCompletionStage<Void> runAfterEither(CompletionStage<?> other, Runnable action)
+  {
+    return produceEitherStage("runAfterEither", cast(other, (v) -> null), (t) -> {
+      action.run();
+      return null;
+    });
+  }
+
+  @Override
+  public ParSeqBasedCompletionStage<Void> runAfterEitherAsync(CompletionStage<?> other, Runnable action,
+      Executor executor)
+  {
+    return produceEitherStageAsync("runAfterEitherAsync", cast(other, (v) -> null), (t) -> {
+      action.run();
+      return null;
+    }, executor);
+  }
+
+  @Override
+  public ParSeqBasedCompletionStage<Void> runAfterEitherAsync(CompletionStage<?> other, Runnable action)
+  {
+    return runAfterEitherAsync(other, action, _asyncExecutor);
+  }
+
+  @Override
+  public ParSeqBasedCompletionStage<T> exceptionally(Function<Throwable, ? extends T> fn)
+  {
+    return nextStageByComposingTask(_task.recover(fn::apply));
+  }
+
+  @Override
+  public <U> ParSeqBasedCompletionStage<U> handle(BiFunction<? super T, Throwable, ? extends U> fn)
+  {
+    return nextStageByComposingTask(_task.transform("handle", prevTaskResult -> {
+      try {
+        return Success.of(fn.apply(prevTaskResult.isFailed() ? null : prevTaskResult.get(), prevTaskResult.getError()));
+      } catch (Throwable throwable) {
+        return Failure.of(throwable);
+      }
+    }));
+  }
+
+  @Override
+  public <U> ParSeqBasedCompletionStage<U> handleAsync(BiFunction<? super T, Throwable, ? extends U> fn,
+      Executor executor)
+  {
+    return nextStageByComposingTask(_task.transformWith("handleAsync", (prevTaskResult) -> Task.blocking(
+        () -> fn.apply(prevTaskResult.isFailed() ? null : prevTaskResult.get(), prevTaskResult.getError()), executor)));
+  }
+
+  @Override
+  public <U> ParSeqBasedCompletionStage<U> handleAsync(BiFunction<? super T, Throwable, ? extends U> fn)
+  {
+    return handleAsync(fn, _asyncExecutor);
+  }
+
+  @Override
+  public ParSeqBasedCompletionStage<T> whenComplete(BiConsumer<? super T, ? super Throwable> action)
+  {
+    return nextStageByComposingTask(_task.transform("whenComplete", prevTaskResult -> {
+      if (prevTaskResult.isFailed()) {
+        try {
+          action.accept(null, prevTaskResult.getError());
+        } catch (Throwable e) {
+          // no ops
+        }
+        return Failure.of(prevTaskResult.getError());
+      } else {
+        try {
+          action.accept(prevTaskResult.get(), prevTaskResult.getError());
+        } catch (Throwable e) {
+          return Failure.of(e);
+        }
+        return Success.of(prevTaskResult.get());
+      }
+    }));
+  }
+
+  @Override
+  public ParSeqBasedCompletionStage<T> whenCompleteAsync(BiConsumer<? super T, ? super Throwable> action,
+      Executor executor)
+  {
+    return nextStageByComposingTask(_task.transformWith("whenCompleteAsync", prevTaskResult -> {
```

Comment:
Calling transform on the `_task` might fail when `_task` is already completed, right? 

Wouldn't it be cleaner to just add a promise listener on _task?

Reply:
actually chaining on a task is equivalent to add a promise listner when the task not resolved. If task resolved, then chained task try to get the value directly.

That also implies that with the promise listner approach, when adding the listern, if the task already finished, added listner won't get notified.

Besides, promise can add listners but task doesn't have this interface. It would be hard to add listner

Reply:
Task extends Promise so you have access to add listener. And promise listener will be notified even if the promise is already fulfilled (check the java doc).

I don't think task chaining is equivalent to add listener. Task chaining will try to run the parent task and can fail if it is already completed (what I figured from looking at code).

Reply:
> Task chaining will try to run the parent task and can fail if it is already completed (what I figured from looking at code)

That should not be the case. (and it would be weird if that is the case)

Say `TaskA` already finished.
```
TaskB = SthNotStarted();
TaskC = TaskA.chainWith(TaskB);
TaskD = TaskA.andThen( val -> Logic() );
```
Do you mean that TaskC and TaskD will fail and TaskB will not run?  I don't udnerstand why. From what I know, TaskA's result (fail or success) will be propagated.

Reply:
btw, the task failure I have previously run through is due to the fact that Travis-CI test machine probably have only 1 core (probably less than 2 CPU ThreadContexts).

In my test, I create two tasks and will complete one with failure after some delay and complete another with success after a longer delay(I used Thread.sleep, which in Java blocks the thread). In my local box it always works but in Travis-CI sometimes it didn't if the longer task starts first (because the shorter one, which supposed to be finished first will run after that)

Reply:
You are correct about chaining on already completed task. However, it might be more performant to use the promise listener? task chaining will schedule the predecessor task to run again. Even if it does nothing, it can still take CPU cycles to execute the runnable etc.
Also we need to see how the trace looks like in both cases. I think task chaining will generate a combined trace. Did you check?

Reply:
(1) I don't see how addListiner is better, except it is harder to implement IMO. I really want to say that before I choose to chain on the task, I carefully considered the add listner option. You will be absoluitely right about it being less performant.  I think it fine and there is good reason to use the task directly: Promise listerns are lower level APIs. Task compositions use higher level APIs(underneath the hood, it uses "After" object, or InternalUtils.after to register the listners). There are already logic embeded inside builtin tasks such as FusionTask or TupleTasks. If we rewrite everything using PromiseListner, it could produce duplicated code, more work and IMO error-prone. It is like how you will choose high level language over hand-crafted asm.

(2) For the trace, I did some tests, looked at the code, and there is some differences if the task already done (but not major difference), so I don't have concerns. Let me explain:
- There are two kinds of relationship: i.e. successor and parental. For task already done, the "parent_of" relationship will become "potential_parent_of", in the trace that's all. The "successor" relationship would remain the same.
- But the main thing is the relationship would not be messed up, because the relationship is defined whenever the context starts a new task(so it is not related to whether the task is done or not). Also the trace is propagated between sucessing tasks only after the previous task is done (this is similar to how value is propagated), so whether the task is done early has no impact on the propagated trace. For details you only need to check [BaseTask](https://github.com/linkedin/parseq/blob/master/subprojects/parseq/src/main/java/com/linkedin/parseq/BaseTask.java) and [FusionTask](https://github.com/linkedin/parseq/blob/master/subprojects/parseq/src/main/java/com/linkedin/parseq/FusionTask.java)

Reply:
Sounds good, thanks for looking into the details. Resolving this conversation.